### ISSUE TITLE
Sunshine Improvement/Suggestion? 

### ISSUE THREAD

**Issue by Misl3d** [Nonary/ResolutionAutomation#4]
With Sunshine you are able to add custom commands, so instead of running the script in the background all the time is it possible to add it as a command to Sunshine so when you connect it starts the script and is only active during that session? 

Thank you!

**Comment by Nonary** [comment]
This task can be accomplished with relative ease, provided one has a basic understanding of computer paths. However, we recognize that not all users may have this knowledge, and as a result, some could find it challenging and potentially struggle with completing the task. To avoid an influx of support questions on Discord regarding setup, we've opted not to include this feature in the base script.

If you are comfortable with using commands on a computer, you can review the code on this branch:
https://github.com/Nonary/ResolutionAutomation/tree/precommand_version

In a nutshell, this version modifies the original script so that it breaks after the stream ends and then launches a new PowerShell process in the background. Since Sunshine initiates pre-commands in the system32 folder, please make sure to input the full path of the script in the Activate.ps1 file, relatives paths would not work.

**Comment by Nonary** [comment]
Once https://github.com/LizardByte/Sunshine/pull/1123 is merged into Sunshine I will plan on adding this feature to this script.

While elevation is not required, it fixes some other issues that would make this script easier to integrate with.

**Comment by Nonary** [comment]
@Misl3d I have a pre-release version of this available for download and testing if you're interested. 

https://github.com/Nonary/ResolutionAutomation/releases/tag/precommand_beta

**Comment by Misl3d** [comment]
Nice! I'll get it and mess around with it asap. Thank you!

**Comment by Misl3d** [comment]
Hey so I got to play around with the new script a little bit. And just things I noticed: 

It worked in changing the resolution when connecting with moonlight but once I disconnected it wouldn't change back. 

Running the function script in console I saw it was grabbing my second monitor settings. 
My display settings were set where my primary monitor was identified as display 2. I fixed it by flipping the cables around. 

This let the script grab the correct monitor resolution. So maybe a config or something to specify which monitor the resolution should grab from.

Even though it was grabbing the correct settings the resolution, it wouldn't change back. I ran out of time so I wasn't able mess around with it more but I thought I should just mention the display identifier issue as it will probably come up in the future. 

I read the AI conversation and it's is crazy how well it works. 

Thank you :)

**Comment by Misl3d** [comment]
After a reboot the script began to pick up my secondary settings again, after trying to find a way to get powershell to output display settings of a specific monitor (without external programs), I failed to do so. So as a band aid, that I actually got working:

I created a display_settings.cfg file in the install folder, with the contents being my default monitor settings:

```
Width: 1920
Height: 1080
Freq: 144
```

Then I just replaced the OnStreamEnd Function with 
```
function OnStreamEnd() {
    $content = Get-Content -Path "$path\display_settings.cfg"

    $width = [int]($content | Select-String -Pattern "Width: (\d+)").Matches.Groups[1].Value
    $height = [int]($content | Select-String -Pattern "Height: (\d+)").Matches.Groups[1].Value
    $freq = [int]($content | Select-String -Pattern "Freq: (\d+)").Matches.Groups[1].Value

	Set-ScreenResolution -Width $width -Height $height -Freq $freq
    break;

}
```

Obviously, this adds another manual setup step but it's a one and done, especially since most people's monitor settings will remain static. So maybe we are overthinking things and setting the client resolution should be dynamic, but reverting can be static? 

So maybe It would be OK to have 2 scripts? 

DO: Read Sunshine logs and change resolution to client resolution
UNDO: Read cfg revert to host resolution

**Comment by Nonary** [comment]
> Hey so I got to play around with the new script a little bit. And just things I noticed:
> 
> It worked in changing the resolution when connecting with moonlight but once I disconnected it wouldn't change back.
> 
> Running the function script in console I saw it was grabbing my second monitor settings. My display settings were set where my primary monitor was identified as display 2. I fixed it by flipping the cables around.
> 
> This let the script grab the correct monitor resolution. So maybe a config or something to specify which monitor the resolution should grab from.
> 
> Even though it was grabbing the correct settings the resolution, it wouldn't change back. I ran out of time so I wasn't able mess around with it more but I thought I should just mention the display identifier issue as it will probably come up in the future.
> 
> I read the AI conversation and it's is crazy how well it works.
> 
> Thank you :)

I have fixed the issues with it not reverting back in the latest release (the file date is newer than the published tag date)

**Comment by Misl3d** [comment]
I tried the latest file and it worked in reverting resolution except for refresh rate, and it looks like its a limit of windows powershell.

I have 2 monitors, the primary being 144hz and second being 60hz, running `Get-CimInstance -ClassName Win32_videocontroller` CurrentRefreshRate outputs 60, I think windows is listing the refresh rate available to all screens. 

So when reverting back to host resolution it will only revert to 60hz and not the 144hz I had it set to.

**Comment by Nonary** [comment]
Ye

> I tried the latest file and it worked in reverting resolution except for refresh rate, and it looks like its a limit of windows powershell.
> 
> I have 2 monitors, the primary being 144hz and second being 60hz, running `Get-CimInstance -ClassName Win32_videocontroller` CurrentRefreshRate outputs 60, I think windows is listing the refresh rate available to all screens.
> 
> So when reverting back to host resolution it will only revert to 60hz and not the 144hz I had it set to.

Yeah this is a problem I've been trying to figure out how to get around. The vast majority of users will have their primary monitor as position 1 instead of 2. And I can't seem to figure out how to use the Windows API to get information such as which is the users primary display; at least not without having the script run as admin which is what I am trying to avoid.

**Comment by Misl3d** [comment]
I have no idea what I am doing inpowershell but here's what I got, I found this : [Reddit Post w/ pastebin in comments](https://www.reddit.com/r/PowerShell/comments/o9qqz6/change_monitor_refresh_rate_strange_need/)

From the [pastebin script ](https://pastebin.com/DupLhevj)I took out the refresh rate grabber and put in ResolutionMatcher.ps1 by replacing the `$hostresolution` line with

```
Add-Type -AssemblyName System.Windows.Forms
$primaryScreen = [System.Windows.Forms.Screen]::PrimaryScreen
$devMode = [Display+DEVMODE]::new()

[Display]::EnumDisplaySettings($primaryScreen.DeviceName,-1,[ref]$devMode)

$hostResolutions = New-Object -TypeName PSObject
$hostResolutions | Add-Member -MemberType NoteProperty -Name CurrentHorizontalResolution -Value $devMode.dmPelsWidth
$hostResolutions | Add-Member -MemberType NoteProperty -Name CurrentVerticalResolution -Value $devMode.dmPelsHeight
$hostResolutions | Add-Member -MemberType NoteProperty -Name CurrentRefreshRate -Value $devMode.dmDisplayFrequency
```

It's messier since idk what I am doing but It worked, it changed my resolution and refresh rate correctly :) 
No admin too!

**Comment by Misl3d** [comment]
It worked until it didn't, Idk what changed but it stopped working and began outputting 0 for settings..

**Comment by Nonary** [comment]
> I have no idea what I am doing inpowershell but here's what I got, I found this : [Reddit Post w/ pastebin in comments](https://www.reddit.com/r/PowerShell/comments/o9qqz6/change_monitor_refresh_rate_strange_need/)
> 
> From the [pastebin script ](https://pastebin.com/DupLhevj)I took out the refresh rate grabber and put in ResolutionMatcher.ps1 by replacing the `$hostresolution` line with
> 
> ```
> Add-Type -AssemblyName System.Windows.Forms
> $primaryScreen = [System.Windows.Forms.Screen]::PrimaryScreen
> $devMode = [Display+DEVMODE]::new()
> 
> [Display]::EnumDisplaySettings($primaryScreen.DeviceName,-1,[ref]$devMode)
> 
> $hostResolutions = New-Object -TypeName PSObject
> $hostResolutions | Add-Member -MemberType NoteProperty -Name CurrentHorizontalResolution -Value $devMode.dmPelsWidth
> $hostResolutions | Add-Member -MemberType NoteProperty -Name CurrentVerticalResolution -Value $devMode.dmPelsHeight
> $hostResolutions | Add-Member -MemberType NoteProperty -Name CurrentRefreshRate -Value $devMode.dmDisplayFrequency
> ```
> 
> It's messier since idk what I am doing but It worked, it changed my resolution and refresh rate correctly :) No admin too!

Windows Forms is an option, but the issue with using DisplayName is that it can actually change once resolution is applied. It's a really frustrating bug in the Windows API. So I can use that but will have to figure out a method that works in the bounds of the bugged Windows API.

**Comment by Nonary** [comment]
@Misl3d I believe I have fixed this issue in the latest release update.

**Comment by Nonary** [comment]
RIP there was a bug with it. It's fixed now.

Also I have refactored the script to no longer rely on the $path being set, so now going forward just simply replacing the script files is enough to update them, no need to run the install each time.

**Comment by Misl3d** [comment]
Nice! I updated and it worked as except both resolution and refresh rate reverted to the correct original values :)

**Comment by Nonary** [comment]
Closing out this issue as it is now implemented as a new release now.

https://github.com/Nonary/ResolutionAutomation/releases/tag/precommand
